lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+            ).all()
+        )
+
+
+@dataclass
+class UnifiedCourse:
+    """All Course's for a given authority_provided_id, unified across all ApplicationInstance's."""
+
+    courses: List[Course] = field(default_factory=list)
+    annotations: List[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        for course in self.courses:
+            assert course.authority_provided_id == self.authority_provided_id
+
+    @property
+    def authority_provided_id(self) -> str:
+        """Return this course's authority_provided_id."""
+        return self.courses[0].authority_provided_id
+
+    @property
+    def title(self) -> str:
+        """Return a title for this course."""
+        return self.courses[0].lms_name
+
+    def learner_annotations(self, db) -> List[dict]:
+        """Return a list of all learner annotations in this course."""
+        return [
+            annotation
+            for annotation in self.annotations
+            if not self.is_instructor(
+                db, UnifiedUser.make(db, annotation["author"]["userid"])
+            )
+        ]
+
+    def is_instructor(self, db, user: UnifiedUser) -> bool:
+        """Return True if `user` is an instructor in this course."""
+        course_ids = [course.id for course in self.courses]
+
+        return bool(
+            db.scalar(
+                select(func.count())
+                .select_from(AssignmentMembership)
+                .join(Assignment)
+                .join(AssignmentGrouping)
+                .join(LTIRole)
+                .filter(AssignmentGrouping.grouping_id.in_(course_ids))
+                .filter(AssignmentMembership.user_id.in_(user.user_ids))
+                .filter(LTIRole.type == "instructor", LTIRole.scope == "course")
+            )
+        )
+
+    @staticmethod
+    def get_courses(db, authority_provided_id):
+        """Return a list of all Course's with the given authority_provided_id."""
+        return db.scalars(
+            select(Course)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Course.updated.desc())
+        ).all()
+
+    @classmethod
+    def make(cls, db, authority_provided_id):
+        """Return a UnifiedCourse for the given authority_provided_id."""
+        first_grouping = db.scalars(
+            select(Grouping)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Grouping.updated.desc())
+        ).first()
+
+        if not first_grouping:
+            raise UnknownAuthorityProvidedID()
+
+        if first_grouping.type == Grouping.Type.COURSE:
+            authority_provided_id = first_grouping.authority_provided_id
+        else:
+            assert first_grouping.parent.type == Grouping.Type.COURSE
+            authority_provided_id = first_grouping.parent.authority_provided_id
+
+        return cls(cls.get_courses(db, authority_provided_id))
```

Comment:
And again: two queries for each group. We might be able to optimise this into one big query to get a mapping of all the course `authority_provided_id`'s to all their groupings. Let's monitor it and see if optimisation is necessary